proposals/p3848.md

```diff
@@ -0,0 +1,576 @@
+# Lambdas
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/3848)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Syntax Overview](#syntax-overview)
+-   [Introducer](#introducer)
+-   [Positional Parameters](#positional-parameters)
+    -   [Positional Parameter Restrictions](#positional-parameter-restrictions)
+-   [Function Captures](#function-captures)
+    -   [Default Capture Mode](#default-capture-mode)
+    -   [Capture Modes](#capture-modes)
+-   [Function Fields](#function-fields)
+-   [Copy Semantics](#copy-semantics)
+-   [Self and Recursion](#self-and-recursion)
+-   [Rationale](#rationale)
+
+<!-- tocstop -->
+
+## Abstract
+
+To support migration from C++ to Carbon, there must be valid syntax to capture
+the behavior of C++ lambdas. They are defined at their point of use and are
+often anonymous, meaning replacing them solely with function declarations will
+create an ergonomic burden compounded by the need for the migration tool to
+select a name. This proposal proposes a path forward to add lambdas to Carbon
+and augment function declarations accordingly.
+
+## Syntax Overview
+
+**Proposal**: A largely continuous syntax between lambdas and function
+declarations.
+
+At a high level, lambdas and function declarations will look like the following.
+
+```
+let lambda1: auto = fn => T.Make();
+
+let lambda2: auto = fn []() -> T { return T.Make(); };
+
+fn FunctionDeclaration1 => T.Make();
+
+fn FunctionDeclaration2[]() -> T { return T.Make(); }
+```
+
+There are functions which return an expression, such that the return type is
+`auto`.
+
+```
+let lambda: auto = fn => T.Make();
+
+fn FunctionDeclaration => T.Make();
+```
+
+And there are functions with an explicit return type that provide a body of
+statements.
+
+```
+let lambda: auto = fn -> T { return T.Make(); };
+
+fn FunctionDeclaration -> T { return T.Make(); }
+```
+
+There are even functions that provide a body of statements but no return value.
+
+```
+let lambda: auto = fn { Print(T.Make()); };
+
+fn FunctionDeclaration { Print(T.Make()); }
+```
+
+Functions support [captures](#function-captures) and [fields](#function-fields)
+in the square brackets.
```

Comment:
```suggestion
in the square brackets (in addition to deduced parameters and the `Self` parameter).
```

Reply:
Worded this in a bit more of a verbose way, only to make it clear that `self: Self` and `addr self: Self*` are specifically for function *declarations* and specifically for such that exist inside a class/interface.